### DISCUSSION TITLE
Providing permanent adresses for users

### DISCUSSION THREAD

**Discussion by nosovk** [btcpayserver/btcpayserver#2763] [Technical Support]
We would like to use BTCPAy as a backend for our wallets solution. Greenfield API seems to be almost perfect solution for a new wallet a case.

In short desired flow is to:
1. Create a shop with wallet that will be store for all coins.
2. Create invoice with UserID obtained from our external application and some very big amount like 9999999BTC. It's done via Greenfiled API.
3. After incoive creation store all addresses connected with that invoice to internal DB as permanent address for wallet top up. On invoice creation addresses for all availble directions automaticaly generated.
4. subscribe on `InvoicePaymentReceived` webhook to get notifiactions about payments to related wallets.

Seems great, but we have a few problems during implementatnion:
1. 34560minutes - is maximum invoice expiration time. I wasn't able to find how to extend it. Is there some way to extend period when transctions are monitored? I tryied extend that time by API, but looks like there is no way to update that param on existing invoice.
2. webhooks are not fired on partial paymetns. Curently we fixed that by making requests for manual update invoce payments using cron. But it's not scalable.

I found some special options in other issues on githib:
```
FullNotifications = true
ExtendedNotifications = true
TransactionSpeed = low
```
it was expected that in such configuration every payment (even partial) to invoice will emit webhook call. But in real life - partial payments does not fire any webhooks.

Would appreciate any ideas how to fix both issues.

**Comment by b40y22** [comment]
I've do similar research, and found that there is `setExpirationTime` field avalible during invoice creation, but it's ignored. And in common we give up on bringing it to work. It looks real timeouts hardcoded somewhere inside.

**Comment by nosovk** [comment]
Similar question: https://github.com/btcpayserver/btcpayserver/discussions/2571

**Comment by Kukks** [comment]
```
FullNotifications = true
ExtendedNotifications = true
```
are for the legacy BitPay API notification system.

`TransactionSpeed` relates only to when a payment is considered `Settled` (which determines the invoice status), not to its actual tracking duration.

`InvoicePaymentReceived` event should be fired on any payment on an invoice, even late/partial payments. There MAY be a situation where the monitoring time has run out and this event is not fired but will need to double-check.


Many years ago, I also attempted this approach but found it undesirable due to the number of edge cases. Fortunately for you, the next release has a great solution for you: [Topup invoices.](https://github.com/btcpayserver/btcpayserver/pull/2730)

**Reply by nosovk** [reply]
We just deploed latest release and testing Topup invoces now. Seems like a great solution.
But even with TopUp invoice address isn't monitored forewer, am I right and BTC adresses will leave monitored state after few weeks?

**Reply by bolatovumar** [reply]
@nosovk yeah, you are right, max expiration time for an invoice is 34560 minutes. I don't know if there is a reason we have this limit in the first place. Looks like @Kukks added this limit in this PR: https://github.com/btcpayserver/btcpayserver/pull/1573

**Reply by nosovk** [reply]
no refference why it was added. Should I create PR to remove that limitations and allow MAX_INT instead?

**Reply by bolatovumar** [reply]
> no refference why it was added. Should I create PR to remove that limitations and allow MAX_INT instead?

That would be a question for @Kukks. There might be a legitimate reason for having that limit.
